Crash : flutter_overlay_window.OverlayService: java.lang.NullPointerException:  flutterChannel & overlayMessageChannel

Hi @X-SLAYER ,

I have facing a Null Pointer Exception in the OverlayService class.

Scenario : 

I have used your flutter_overlay_window package, and I have a home screen widget, when I will click on that home screen widget this flutter_overlay_window should open.

Error:
When I have installed the app, and given the required permissions, and placed the widget on the home screen. Once I clicked the widget I was facing this crash.

Logs:

```
E  FATAL EXCEPTION: main
Process: com.example.flutter_overlay_window_example, PID: 32151
java.lang.RuntimeException: Unable to create service flutter.overlay.window.flutter_overlay_window.OverlayService: java.lang.NullPointerException: Attempt to invoke virtual method 'w.a io.flutter.embedding.engine.a.j()' on a null object reference
at android.app.ActivityThread.handleCreateService(ActivityThread.java:4925)
at android.app.ActivityThread.-$$Nest$mhandleCreateService(Unknown Source:0)
at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2407)
at android.os.Handler.dispatchMessage(Handler.java:107)
at android.os.Looper.loopOnce(Looper.java:232)
at android.os.Looper.loop(Looper.java:317)
at android.app.ActivityThread.main(ActivityThread.java:8592)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:580)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:878)
Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'w.a io.flutter.embedding.engine.a.j()' on a null object reference
at flutter.overlay.window.flutter_overlay_window.OverlayService.<init>(Unknown Source:35)
at java.lang.Class.newInstance(Native Method)
at android.app.AppComponentFactory.instantiateService(AppComponentFactory.java:129)
at androidx.core.app.CoreComponentFactory.instantiateService(Unknown Source:0)
at android.app.ActivityThread.handleCreateService(ActivityThread.java:4884)
at android.app.ActivityThread.-$$Nest$mhandleCreateService(Unknown Source:0) 
at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2407) 
at android.os.Handler.dispatchMessage(Handler.java:107) 
at android.os.Looper.loopOnce(Looper.java:232) 
at android.os.Looper.loop(Looper.java:317) 
at android.app.ActivityThread.main(ActivityThread.java:8592) 
at java.lang.reflect.Method.invoke(Native Method) 
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:580) 
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:878) 

```


![Image](https://github.com/user-attachments/assets/e4bae94a-1db9-40fd-ae48-6564a0c5956d)

After investigating a bit further, I got to know that at line :

https://github.com/X-SLAYER/flutter_overlay_window/blob/faaff1f965649b0410c8107745bb64a8ff86a687/android/src/main/java/flutter/overlay/window/flutter_overlay_window/OverlayService.java#L58

and

https://github.com/X-SLAYER/flutter_overlay_window/blob/faaff1f965649b0410c8107745bb64a8ff86a687/android/src/main/java/flutter/overlay/window/flutter_overlay_window/OverlayService.java#L59


You are trying to access FlutterEngine `FlutterEngineCache.getInstance().get(OverlayConstants.CACHED_TAG).getDartExecutor()`. 

But in my case as I didn't even opened the app, and trying to access the overlay, it's crashing the app with this `NullPointerException'

### Possible Fix 💯 
We need to initialize `flutterChannel` and `overlayMessageChannel` in the onCreate method. We need to add have a condition in the `onCreate` method to check if the flutter engine is null or not. If null we need to create the flutter engine and then we need to create the method channel.



@X-SLAYER  I would like to raise a PR for this FIX. You can review it.

--- a/android/src/main/java/flutter/overlay/window/flutter_overlay_window/OverlayService.java
+++ b/android/src/main/java/flutter/overlay/window/flutter_overlay_window/OverlayService.java
@@ -35,8 +35,11 @@ import java.util.TimerTask;
 
 import io.flutter.embedding.android.FlutterTextureView;
 import io.flutter.embedding.android.FlutterView;
+import io.flutter.FlutterInjector;
 import io.flutter.embedding.engine.FlutterEngine;
 import io.flutter.embedding.engine.FlutterEngineCache;
+import io.flutter.embedding.engine.FlutterEngineGroup;
+import io.flutter.embedding.engine.dart.DartExecutor;
 import io.flutter.plugin.common.BasicMessageChannel;
 import io.flutter.plugin.common.JSONMessageCodec;
 import io.flutter.plugin.common.MethodChannel;
@@ -55,8 +58,8 @@ public class OverlayService extends Service implements View.OnTouchListener {
     public static boolean isRunning = false;
     private WindowManager windowManager = null;
     private FlutterView flutterView;
-    private MethodChannel flutterChannel = new MethodChannel(FlutterEngineCache.getInstance().get(OverlayConstants.CACHED_TAG).getDartExecutor(), OverlayConstants.OVERLAY_TAG);
-    private BasicMessageChannel<Object> overlayMessageChannel = new BasicMessageChannel(FlutterEngineCache.getInstance().get(OverlayConstants.CACHED_TAG).getDartExecutor(), OverlayConstants.MESSENGER_TAG, JSONMessageCodec.INSTANCE);
+    private MethodChannel flutterChannel;
+    private BasicMessageChannel<Object> overlayMessageChannel;
     private int clickableFlag = WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE | WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE |
             WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS | WindowManager.LayoutParams.FLAG_LAYOUT_IN_SCREEN;
 
@@ -295,6 +298,30 @@ public class OverlayService extends Service implements View.OnTouchListener {
 
     @Override
     public void onCreate() {
+        // Get the cached FlutterEngine
+        FlutterEngine flutterEngine = FlutterEngineCache.getInstance().get(OverlayConstants.CACHED_TAG);
+
+        if (flutterEngine == null) {
+            // Handle the error if engine is not found
+            Log.e("OverlayService", "Flutter engine not found, hence creating new flutter engine");
+            FlutterEngineGroup engineGroup = new FlutterEngineGroup(this);
+            DartExecutor.DartEntrypoint entryPoint = new DartExecutor.DartEntrypoint(
+                FlutterInjector.instance().flutterLoader().findAppBundlePath(),
+                "overlayMain"
+            );  // "overlayMain" is custom entry point
+
+            flutterEngine = engineGroup.createAndRunEngine(this, entryPoint);
+
+            // Cache the created FlutterEngine for future use
+            FlutterEngineCache.getInstance().put(OverlayConstants.CACHED_TAG, flutterEngine);
+        }
+
+        // Create the MethodChannel with the properly initialized FlutterEngine
+        if (flutterEngine != null) {
+            flutterChannel = new MethodChannel(flutterEngine.getDartExecutor(), OverlayConstants.OVERLAY_TAG);
+            overlayMessageChannel = new BasicMessageChannel(flutterEngine.getDartExecutor(), OverlayConstants.MESSENGER_TAG, JSONMessageCodec.INSTANCE);
+        }
+
         createNotificationChannel();
         Intent notificationIntent = new Intent(this, FlutterOverlayWindowPlugin.class);
         int pendingFlags;

